File: app/Klsf/Dns/West.php
<?php

namespace App\Klsf\Dns;

class West implements DnsInterface
{
    use DnsHttp;

    private $url = 'https://api.west.cn/api/v2/domain/';
    private $username;
    private $password;

    public function deleteDomainRecord($RecordId, $DomainId, $Domain)
    {
        list($ret, $error) = $this->getResult('deldnsrecord', [
            'domain' => $Domain,
            'id' => $RecordId,
        ]);

        return $ret ? [true, null] : [false, $error];
    }

    public function updateDomainRecord($RecordId, $Name, $Type, $Value, $LineId, $DomainId = null, $Domain = null)
    {
        list($ret, $error) = $this->getResult('moddnsrecord', [
            'domain' => $Domain,
            'id' => $RecordId,
            'host' => $this->normalizeHost($Name),
            'type' => strtoupper(trim((string) $Type)),
            'value' => trim((string) $Value),
            'line' => $this->normalizeLine($LineId),
            'ttl' => 900,
            'level' => $this->normalizeLevel($Type),
        ]);

        return $ret ? [true, null] : [false, $error];
    }

    public function addDomainRecord($Name, $Type, $Value, $LineId, $DomainId = null, $Domain = null)
    {
        list($ret, $error) = $this->getResult('adddnsrecord', [
            'domain' => $Domain,
            'host' => $this->normalizeHost($Name),
            'type' => strtoupper(trim((string) $Type)),
            'value' => trim((string) $Value),
            'line' => $this->normalizeLine($LineId),
            'ttl' => 900,
            'level' => $this->normalizeLevel($Type),
        ]);

        if (!$ret) {
            return [false, $error];
        }

        if (isset($ret['data']['id'])) {
            return [[
                'RecordId' => $ret['data']['id'],
                'Name' => $Name,
                'Domain' => $Domain,
            ], null];
        }

        return [false, '添加域名记录失败'];
    }

    public function getDomainRecordInfo($RecordId, $DomainId = null, $Domain = null)
    {
        list($list, $error) = $this->getDomainRecords($DomainId, $Domain);
        if ($list === false) {
            return [false, $error];
        }

        foreach ($list as $record) {
            if ((string) $record['RecordId'] === (string) $RecordId) {
            return [[$record], null];
        }
        }

        return [false, '获取域名记录详情失败'];
    }

    public function getDomainRecords($DomainId = null, $Domain = null)
    {
        list($ret, $error) = $this->getResult('getdnsrecord', [
            'domain' => $Domain,
            'limit' => 500,
            'pageno' => 1,
        ]);

        if (!$ret) {
            return [false, $error];
        }

        $items = isset($ret['data']['items']) && is_array($ret['data']['items']) ? $ret['data']['items'] : [];
        $list = [];

        foreach ($items as $record) {
            $list[] = [
                'RecordId' => $record['id'],
                'Name' => $this->denormalizeHost($record['item'] ?? '@'),
                'Type' => strtoupper((string) ($record['type'] ?? 'A')),
                'Value' => (string) ($record['value'] ?? ''),
                'Domain' => $Domain,
            ];
        }

        return [$list, null];
    }

    public function getDomainList()
    {
        list($ret, $error) = $this->getResult('getdomains', [
            'limit' => 500,
            'page' => 1,
        ], 'GET');

        if (!$ret) {
            return [false, $error];
        }

        $items = isset($ret['data']['items']) && is_array($ret['data']['items']) ? $ret['data']['items'] : [];
        $list = [];

        foreach ($items as $domain) {
            if (!empty($domain['domain'])) {
                $list[] = [
                    'Domain' => $domain['domain'],
                    'DomainId' => $domain['domain'],
                ];
            }
        }

        return [$list, null];
    }

    public function getRecordLine($_domainId = null, $_domain = null)
    {
        return [
            ['Name' => '默认', 'Id' => ''],
            ['Name' => '电信', 'Id' => 'LTEL'],
            ['Name' => '联通', 'Id' => 'LCNC'],
            ['Name' => '移动', 'Id' => 'LMOB'],
            ['Name' => '教育网', 'Id' => 'LEDU'],
            ['Name' => '搜索引擎', 'Id' => 'LSEO'],
        ];
    }

    public function check()
    {
        list($ret, $error) = $this->getDomainList();

        return $ret ? [true, null] : [false, $error];
    }

    public function config(array $config)
    {
        $this->username = isset($config['Username']) ? trim((string) $config['Username']) : '';
        $this->password = isset($config['ApiPassword']) ? trim((string) $config['ApiPassword']) : '';
    }

    public function configInfo()
    {
        return [
            [
                'name' => 'Username',
                'placeholder' => '请输入西部数码 API 用户名',
                'tips' => '<a href="https://www.west.cn/CustomerCenter/doc/domain_v2.html" target="_blank">查看西部数码域名 API 文档</a>',
            ],
            [
                'name' => 'ApiPassword',
                'placeholder' => '请输入 API 密码',
                'tips' => '请使用西部数码开放的 API 密码，不是登录后台密码。',
            ],
        ];
    }

    private function getResult($action, $params = [], $method = 'POST')
    {
        $timestamp = (string) round(microtime(true) * 1000);
        $query = array_merge([
            'username' => $this->username,
            'time' => $timestamp,
            'token' => md5($this->username . $this->password . $timestamp),
            'act' => $action,
        ], $params);

        if (strtoupper($method) === 'GET') {
            list($res, $error) = $this->get($this->url . '?' . http_build_query($query));
        } else {
            list($res, $error) = $this->post($this->url, [
                'body' => http_build_query($query),
            ]);
        }

        if (!$res) {
            return [false, $error];
        }

        $body = (string) $res->getBody();
        $body = $this->normalizeBody($body);
        $ret = json_decode($body, true);

        if (!is_array($ret)) {
            return [false, '解析结果失败'];
        }

        if (isset($ret['result']) && intval($ret['result']) === 200) {
            return [$ret, null];
        }

        return [false, $ret['msg'] ?? '接口调用失败'];
    }

    private function normalizeBody($body)
    {
        if (!is_string($body) || $body === '') {
            return '';
        }

        $encoding = mb_detect_encoding($body, ['UTF-8', 'GBK', 'GB2312', 'BIG5'], true);
        if ($encoding && strtoupper($encoding) !== 'UTF-8') {
            return mb_convert_encoding($body, 'UTF-8', $encoding);
        }

        return $body;
    }

    private function normalizeHost($host)
    {
        $host = trim((string) $host);

        return $host === '' ? '@' : $host;
    }

    private function denormalizeHost($host)
    {
        $host = trim((string) $host);

        return $host === '@' ? '' : $host;
    }

    private function normalizeLine($lineId)
    {
        $lineId = trim((string) $lineId);

        return $lineId === '0' ? '' : $lineId;
    }

    private function normalizeLevel($type)
    {
        return strtoupper(trim((string) $type)) === 'MX' ? 10 : 10;
    }
}

File: pkg/dns/providers/west.go
package providers

import (
	"context"
	"crypto/md5"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strconv"
	"strings"
	"time"

	"kldns/pkg/dns"
	"kldns/pkg/dns/providerhttp"
)

const westDefaultBaseURL = "https://api.west.cn/api/v2/domain/"

func init() {
	dns.Register("West", func() dns.Provider {
		return &westProvider{
			baseURL: westDefaultBaseURL,
			client:  providerhttp.NewClient(),
			nowMS:   func() int64 { return time.Now().UnixMilli() },
		}
	})
}

type westProvider struct {
	username string
	password string
	baseURL  string
	client   *http.Client
	nowMS    func() int64
}

type westResponse struct {
	Result int      `json:"result"`
	Msg    string   `json:"msg"`
	Data   westData `json:"data"`
}

type westData struct {
	ID    json.RawMessage `json:"id"`
	Items []westItem      `json:"items"`
}

type westItem struct {
	ID     json.RawMessage `json:"id"`
	Domain string          `json:"domain"`
	Item   string          `json:"item"`
	Type   string          `json:"type"`
	Value  string          `json:"value"`
	Line   string          `json:"line"`
}

func (p *westProvider) Key() string {
	return "West"
}

func (p *westProvider) Label() string {
	return "西部数码"
}

func (p *westProvider) ConfigFields() []dns.ConfigField {
	return []dns.ConfigField{
		{Name: "Username", Label: "Username", Required: true},
		{Name: "ApiPassword", Label: "ApiPassword", Required: true, Secret: true},
	}
}

func (p *westProvider) Configure(config map[string]string) error {
	p.username = strings.TrimSpace(config["Username"])
	p.password = strings.TrimSpace(config["ApiPassword"])
	if baseURL := strings.TrimSpace(config["BaseURL"]); baseURL != "" {
		p.baseURL = baseURL
	}
	if p.baseURL == "" {
		p.baseURL = westDefaultBaseURL
	}
	if p.client == nil {
		p.client = providerhttp.NewClient()
	}
	if p.nowMS == nil {
		p.nowMS = func() int64 { return time.Now().UnixMilli() }
	}
	return nil
}

func (p *westProvider) Check(ctx context.Context) error {
	_, err := p.ListZones(ctx)
	return err
}

func (p *westProvider) ListZones(ctx context.Context) ([]dns.Zone, error) {
	ret, err := p.doRequest(ctx, http.MethodGet, "getdomains", map[string]string{"limit": "500", "page": "1"}, "list_zones")
	if err != nil {
		return nil, err
	}
	zones := make([]dns.Zone, 0, len(ret.Data.Items))
	for _, item := range ret.Data.Items {
		if strings.TrimSpace(item.Domain) != "" {
			zones = append(zones, dns.Zone{ID: item.Domain, Domain: item.Domain})
		}
	}
	return zones, nil
}

func (p *westProvider) ListRecordLines(context.Context, dns.Zone) ([]dns.RecordLine, error) {
	return []dns.RecordLine{
		{ID: "", Name: "默认"},
		{ID: "LTEL", Name: "电信"},
		{ID: "LCNC", Name: "联通"},
		{ID: "LMOB", Name: "移动"},
		{ID: "LEDU", Name: "教育网"},
		{ID: "LSEO", Name: "搜索引擎"},
	}, nil
}

func (p *westProvider) CreateRecord(ctx context.Context, zone dns.Zone, input dns.RecordInput) (dns.Record, error) {
	params := westRecordParams(zone, input)
	ret, err := p.doRequest(ctx, http.MethodPost, "adddnsrecord", params, "create_record")
	if err != nil {
		return dns.Record{}, err
	}
	result := dns.Record{
		RemoteID: rawString(ret.Data.ID),
		Name:     denormalizeWestHost(params["host"]),
		Type:     params["type"],
		Value:    params["value"],
		LineID:   params["line"],
		Line:     westLineName(params["line"]),
	}
	if result.RemoteID == "" {
		return dns.Record{}, &dns.ProviderError{Provider: p.Key(), Operation: "create_record", Message: "missing record id in response"}
	}
	return result, nil
}

func (p *westProvider) UpdateRecord(ctx context.Context, zone dns.Zone, remoteID string, input dns.RecordInput) (dns.Record, error) {
	params := westRecordParams(zone, input)
	params["id"] = strings.TrimSpace(remoteID)
	if _, err := p.doRequest(ctx, http.MethodPost, "moddnsrecord", params, "update_record"); err != nil {
		return dns.Record{}, err
	}
	return dns.Record{
		RemoteID: remoteID,
		Name:     denormalizeWestHost(params["host"]),
		Type:     params["type"],
		Value:    params["value"],
		LineID:   params["line"],
		Line:     westLineName(params["line"]),
	}, nil
}

func (p *westProvider) DeleteRecord(ctx context.Context, zone dns.Zone, remoteID string) error {
	_, err := p.doRequest(ctx, http.MethodPost, "deldnsrecord", map[string]string{
		"domain": zone.Domain,
		"id":     strings.TrimSpace(remoteID),
	}, "delete_record")
	return err
}

func (p *westProvider) GetRecord(ctx context.Context, zone dns.Zone, remoteID string) (dns.Record, error) {
	records, err := p.ListRecords(ctx, zone)
	if err != nil {
		return dns.Record{}, err
	}
	for _, record := range records {
		if record.RemoteID == remoteID {
			return record, nil
		}
	}
	return dns.Record{}, &dns.ProviderError{Provider: p.Key(), Operation: "get_record", Message: "record not found"}
}

func (p *westProvider) ListRecords(ctx context.Context, zone dns.Zone) ([]dns.Record, error) {
	ret, err := p.doRequest(ctx, http.MethodPost, "getdnsrecord", map[string]string{
		"domain": zone.Domain,
		"limit":  "500",
		"pageno": "1",
	}, "list_records")
	if err != nil {
		return nil, err
	}
	records := make([]dns.Record, 0, len(ret.Data.Items))
	for _, item := range ret.Data.Items {
		line := normalizeWestLine(item.Line)
		records = append(records, dns.Record{
			RemoteID: rawString(item.ID),
			Name:     denormalizeWestHost(item.Item),
			Type:     strings.ToUpper(strings.TrimSpace(item.Type)),
			Value:    item.Value,
			LineID:   line,
			Line:     westLineName(line),
		})
	}
	return records, nil
}

func (p *westProvider) doRequest(ctx context.Context, method string, action string, params map[string]string, operation string) (westResponse, error) {
	if err := p.validateAuth(operation); err != nil {
		return westResponse{}, err
	}
	values := url.Values{}
	timestamp := strconv.FormatInt(p.nowMS(), 10)
	values.Set("username", p.username)
	values.Set("time", timestamp)
	values.Set("token", westToken(p.username, p.password, timestamp))
	values.Set("act", action)
	for key, value := range params {
		values.Set(key, value)
	}
	endpoint := p.baseURL
	var body io.Reader
	if method == http.MethodGet {
		sep := "?"
		if strings.Contains(endpoint, "?") {
			sep = "&"
		}
		endpoint += sep + values.Encode()
	} else {
		body = strings.NewReader(values.Encode())
	}
	req, err := http.NewRequestWithContext(ctx, method, endpoint, body)
	if err != nil {
		return westResponse{}, &dns.ProviderError{Provider: p.Key(), Operation: operation, Message: "create request failed", Cause: err}
	}
	if method != http.MethodGet {
		req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
	}
	resp, err := p.client.Do(req)
	if err != nil {
		return westResponse{}, &dns.ProviderError{Provider: p.Key(), Operation: operation, Message: "request failed", Cause: err}
	}
	defer resp.Body.Close()
	data, err := io.ReadAll(io.LimitReader(resp.Body, 4<<20))
	if err != nil {
		return westResponse{}, &dns.ProviderError{Provider: p.Key(), Operation: operation, Message: "read response failed", Cause: err}
	}
	var ret westResponse
	if err := json.Unmarshal(data, &ret); err != nil {
		return westResponse{}, &dns.ProviderError{Provider: p.Key(), Operation: operation, Message: "decode response failed"}
	}
	if resp.StatusCode < 200 || resp.StatusCode >= 300 || ret.Result != 200 {
		return westResponse{}, &dns.ProviderError{Provider: p.Key(), Operation: operation, Message: westErrorMessage(resp.StatusCode, ret.Msg)}
	}
	return ret, nil
}

func (p *westProvider) validateAuth(operation string) error {
	if p.username != "" && p.password != "" {
		return nil
	}
	return &dns.ProviderError{Provider: p.Key(), Operation: operation, Message: "missing Username or ApiPassword"}
}

func westRecordParams(zone dns.Zone, input dns.RecordInput) map[string]string {
	recordType := strings.ToUpper(strings.TrimSpace(input.Type))
	return map[string]string{
		"domain": zone.Domain,
		"host":   normalizeWestHost(input.Name),
		"type":   recordType,
		"value":  strings.TrimSpace(input.Value),
		"line":   normalizeWestLine(input.LineID),
		"ttl":    "900",
		"level":  westLevel(recordType),
	}
}

func westToken(username string, password string, timestamp string) string {
	sum := md5.Sum([]byte(username + password + timestamp))
	return hex.EncodeToString(sum[:])
}

func normalizeWestHost(host string) string {
	host = strings.TrimSpace(host)
	if host == "" {
		return "@"
	}
	return host
}

func denormalizeWestHost(host string) string {
	host = strings.TrimSpace(host)
	if host == "@" {
		return ""
	}
	return host
}

func normalizeWestLine(line string) string {
	line = strings.TrimSpace(line)
	if line == "0" {
		return ""
	}
	return line
}

func westLineName(line string) string {
	switch normalizeWestLine(line) {
	case "":
		return "默认"
	case "LTEL":
		return "电信"
	case "LCNC":
		return "联通"
	case "LMOB":
		return "移动"
	case "LEDU":
		return "教育网"
	case "LSEO":
		return "搜索引擎"
	default:
		return line
	}
}

func westLevel(recordType string) string {
	if strings.ToUpper(strings.TrimSpace(recordType)) == "MX" {
		return "10"
	}
	return "10"
}

func westErrorMessage(status int, message string) string {
	message = strings.TrimSpace(message)
	if message != "" {
		return message
	}
	if status > 0 {
		return fmt.Sprintf("West API returned HTTP %d", status)
	}
	return "West API request failed"
}
